src/dune_rules/lib.ml

```diff
@@ -1674,12 +1683,34 @@ let descriptive_closure (l : lib list) ~with_pps : lib list Memo.t =
   (* and then convert it to a list *)
   Set.to_list trans_closure
 
+let unique_sublists lsts =
+  let rec aux seen = function
+    | [] -> []
+    | h :: t ->
+      let new_list = List.filter ~f:(fun x -> not (Set.mem seen x)) h in
+      if List.is_empty new_list then new_list :: aux seen t
+      else
+        new_list
+        :: aux
+             (List.fold_right new_list ~init:seen ~f:(fun a b -> Set.add b a))
```

Comment:
I randomly happened to look at the PR, and found this spot. Any reason not to use a tailrec List.fold_left here?

Performance wise, it could also be the case that using a Hashtbl to collect "seen" objects would work better here (I've no idea whether this is a bottleneck!).  You could also filter out already-seen values and add objects to the table with a single pass on the `h` list.